### DISCUSSION TITLE
rqlite on Nix

### DISCUSSION THREAD

**Discussion by siiky** [rqlite/rqlite#1012] [General]
Hullo o/

I want to start using rqlite for a project of mine, and started by trying to build it with Nix. 

According to some knowledgeable Nix folks, Nix builds execute in a [sandbox with no network interface](https://discourse.nixos.org/t/help-building-rqlite-with-buildgomodule/17979/2), which resulted in the tests failing.

To work around this, just to get something going, I disabled tests completely (with `doCheck = false`), and now I have working executables!

```sh
$ /result/bin/rqlite --version
Version 7, commit unknown, branch unknown, built on unknown
$ ./result/bin/rqlited --version
rqlited 7 linux amd64 go1.17.7 (commit unknown, branch unknown, compiler gc)
```

However, I would like to have some peace of mind and see some tests passing. I don't know if all of them need a network interface to run, but that is what I would like to find out with this ticket.

Is there a way to disable specific tests, or something to make it possible to run the tests in a sandboxed environment? Is it worth while to disable specific tests? How many of all the tests actually need the network interface?

---

This is the `default.nix` file I got working:

```nix
{ buildGoModule, fetchFromGitHub, lib }:
buildGoModule rec {
  name = "rqlite";
  version = "7.3.2";
  src = fetchFromGitHub {
    owner = "rqlite";
    repo = "rqlite";
    rev = "v${version}";
    sha256 = "sha256-C0S+zOCdWnUcbrpQeV4dK1n67GrAiYOJJv7VKoEQgUs=";
  };

  doCheck = false;
  runVend = false;
  vendorSha256 = "sha256-Kr1Ug3zuSCeBucz0zpcouYfU0xKGmi0wi9Kh8+adHD4=";

  meta = with lib; {
    description = "The lightweight, distributed relational database built on SQLite";
    homepage = "http://www.rqlite.io";
    license = licenses.mit;
    platforms = platforms.linux ++ platforms.darwin;
  };
}
```

And I use it with `nix-build -E 'with import <nixpkgs> {}; callPackage ./default.nix {}'`. 

Eventually I would like to start building from a locally checked out Git branch, but for now it fetches the release 7.3.2 from GH and builds that, along with all of the dependencies.

**Comment by otoolep** [comment]
Just to be clear, simply building rqlite (`go build`) doesn't trigger any tests, so I'm not sure how that is related. However the Go tool chain will attempt to download packages it needs to build rqlite, so it's not clear how you can build without network connectivity.

Raft and rqlite are fundamentally networking-centric systems. rqlite in particular has always been designed and implemented with the assumption there is a network (even a single `localhost` interface suffices thought). One could have, in principle, used a networking abstraction that used, say, unix sockets underneath, but that would have been a large change. The existence of a TCP networking stack is pretty built into the code, from top-to-bottom, both for the production code and the test framework.  There is no way to easily disable all tests that don't require the network, except going each test and examining them by hand.

**Reply by siiky** [reply]
Thanks for your really quick reply!

> However the Go tool chain will attempt to download packages it needs to build rqlite, so it's not clear how you can build without network connectivity.

Nix downloads all sources first, and only then it builds everything in the sandbox.

---

Indeed, it's a fair assumption, just wanted to be sure. I'll keep skipping all tests then.

**Comment by otoolep** [comment]
In case you didn't see it, official build instructions are here: https://github.com/rqlite/rqlite/blob/master/CONTRIBUTING.md#building-rqlite

You can see how production releases are built by checking out this script: https://github.com/rqlite/rqlite/blob/master/package.sh

**Reply by siiky** [reply]
Yeah, I had seen and read the build instructions, whence I could see that there's nothing fancy going on. The `package.sh` I hadn't seen yet though, thanks! I'll read it and see if there's something of interest that I can adapt for Nix.
